### ISSUE TITLE
Bug on https://www.amazon.com

### ISSUE THREAD

**Issue by Ajumeaux** [Richienb/url-exist#36]
It returns false on this website

**Comment by Richienb** [comment]
That url returns a HTTP 503:

![A HTTP HEAD request to amazon.com which returns a HTTP 503 error](https://user-images.githubusercontent.com/29491356/133760835-ab10a3cb-0730-423b-a0d9-6ddd69a5ded5.png)

**Comment by Richienb** [comment]
The solution might be to also send a GET request if a HEAD request failed but this might need to be an option since most of the time, it would just be sending a second request and not get a different result.

**Comment by Ajumeaux** [comment]
Yes you're right, it would slow the process a bit, but double checking with GET request after an HEAD one could be a nice option

**Comment by Richienb** [comment]
What if we just sent a get request and abort it as soon as a HTTP code was received but before any data is received?

**Comment by Ajumeaux** [comment]
I didn't think of it that way, good idea it would keep the package simple to use, while not really slowing the process

**Comment by Ajumeaux** [comment]
It seems like we can't really abort a get request, I found a lead with XMLHttpRequest.abort(), wich aborts a request that has already been sent. can't really use ky.get() for this or i must be missing something. Doing the head request then the get request (or the get one first) is the simplest option, but it would greatly slow the process. Another option it to call ky.get and ky.head at the same time with promise.all, but we would have to wait for both of them to be complete. I haven't found the third option yet, idk how to abort a get request

**Comment by Ajumeaux** [comment]
```
import isUrl from "is-url-superb"
import ky from "ky-universal"

const urlExist = async url => {
	if (typeof url !== "string") {
		throw new TypeError(`Expected a string, got ${typeof url}`)
	}

	if (!isUrl(url)) {
		return false
	}

	const response = await ky.head(url, {
		throwHttpErrors: false
	})

	if (response !== undefined && (response.status < 400 || response.status >= 500)) {
		return true
	} else {
		const req = await ky.get(url, {
			throwHttpErrors: false
		})
		return req !== undefined && (req.status < 400 || req.status >= 500)
	}
}

export default urlExist
```
Not very elegant, but functional. Slower when head method is not allowed by the host (that's the case for amazon) or when the url does no exists. but I found that the status for method not allowed is 405, so we can get in the else block only if the status is 405, that would mean that we can't use the header method for this url and we should try a get. but if the status is anything but 405 we would return false

**Comment by alonushhh** [comment]
If a URL results in a 403 (forbidden),  a http timeout (firewalls) or non 200 OK this npm returns false. Is there any possibility to add a http status code value so people can inspect what the status code was as well?

**Comment by Ajumeaux** [comment]
actually if an URL is forbidden doesn't it means it exists ?

**Comment by Richienb** [comment]
Yes, that's a bug.

**Comment by jdwillemse** [comment]
> Another option it to call ky.get and ky.head at the same time with promise.all, but we would have to wait for both of them to be complete. I haven't found the third option yet, idk how to abort a get request

Using `Promise.any` can be used here. Run both `ky.head` and `ky.get` and validate the individual responses. If the head response is valid the `Promise.any` is resolved.

```js
const validateResponse = (response) =>
  response !== undefined && (response.status < 400 || response.status >= 500)
    ? true
    : Promise.reject()

const response = await Promise.any([
  ky
    .head(url, {
      throwHttpErrors: false,
    })
    .then(validateResponse),
  ky
    .get(url, {
      throwHttpErrors: false,
    })
    .then(validateResponse),
]).catch(() => false)

return response
```
